src/promptflow-devkit/promptflow/_sdk/_utilities/chat_utils.py

```diff
@@ -10,12 +28,121 @@ def construct_flow_absolute_path(flow: str) -> str:
 
 
 # Todo: use base64 encode for now, will consider whether need use encryption or use db to store flow path info
-def construct_chat_page_url(flow, port, url_params, enable_internal_features=False):
-    flow_path_dir, flow_path_file = resolve_flow_path(flow)
-    flow_path = str(flow_path_dir / flow_path_file)
+def construct_chat_page_url(flow_path, port, url_params):
     encrypted_flow_path = encrypt_flow_path(flow_path)
-    query_dict = {"flow": encrypted_flow_path}
-    if enable_internal_features:
-        query_dict.update({"enable_internal_features": "true", **url_params})
+    query_dict = {"flow": encrypted_flow_path, **url_params}
     query_params = urlencode(query_dict)
     return urlunparse(("http", f"127.0.0.1:{port}", "/v1.0/ui/chat", "", query_params, ""))
+
+
+def _try_restart_service(
+    *, last_result: ServeAppHelper, flow_file_name: str, flow_dir: Path, serve_app_port: int, ux_input_path: Path
+):
+    if last_result is not None:
+        print("Changes detected, stopping current serve app...")
+        last_result.terminate()
+
+    # init must be always loaded from ux_inputs.json
+    if not ux_input_path.is_file():
+        init = {}
+    else:
+        ux_inputs = json.loads(ux_input_path.read_text(encoding=DEFAULT_ENCODING))
+        init = ux_inputs.get(UX_INPUTS_INIT_KEY, {}).get(flow_file_name, {})
+
+    language = resolve_flow_language(flow_path=flow_file_name, working_dir=flow_dir)
+    if language == FlowLanguage.Python:
+        # we should always resolve additional includes as user will change files in original locations only
+        flow_dir = _resolve_python_flow_additional_includes(flow_file_name, flow_dir)
+
+        helper = PythonServeAppHelper(
+            flow_file_name=flow_file_name,
+            flow_dir=flow_dir,
+            init=init,
+            port=int(serve_app_port),
```

Comment:
why do extra int() here?

Reply:
it's not necessary. haven't remove this when refactoring